src/utilities.lisp

```diff
@@ -167,36 +167,34 @@ The result will be a list of cons cells representing half-open intervals (on the
   #-unix
   1)
 
-(let ((prepared? nil)
-      (workers-allotted nil))
-  (defun prepare-for-parallelization (&optional num-workers)
-    "Prepare for parallelization with the correct number of workers scaling with the number of logical cores of your machine.
+(defun prepare-for-parallelization (&optional num-workers)
+  "Create a worker pool if none exists or if NUM-WORKERS has changed.
 
-If NUM-WORKERS is provided, it can force the number of workers. If it's greater than 1, then it should be less than the number of logical cores of your machine.
+If NUM-WORKERS is not provided, the number of workers will be set to the number of logical cores of your machine. ~
+This function does nothing if NUM-WORKERS workers have already been created. ~
+If NUM-WORKERS is provided it should be less than the number of logical cores of your machine.
 
 NOTE: This must be done before computations can be done.
 "
-    (check-type num-workers (or null (integer 1)))
-    (let ((num-logical-cores (or num-workers (count-logical-cores))))
-      (assert (or (null workers-allotted)
-                  (= 1 workers-allotted)
-                  (<= workers-allotted num-logical-cores))
-              ()
-              "The number of workers for parallelization exceeds the ~
-               number of cores. This could be because ~
-               #'QVM:PREPARE-FOR-INITIALIZATION was called too early. ~
-               The number of workers is ~D and the number of logical ~
-               cores is ~D."
-              workers-allotted
-              num-logical-cores)
-      (unless prepared?
-        (let ((num-workers (max 1 num-logical-cores)))
-          (setf lparallel:*kernel*
-                (lparallel:make-kernel num-workers :name "QVM Worker"))
-          (setf workers-allotted num-workers)
-          (setf prepared? t))))
-
-    (values)))
+  (check-type num-workers (or null (integer 1)))
+  (assert (or (null num-workers) (<= num-workers (count-logical-cores)))
+          ()
+          "The number of workers for parallelization exceeds the ~
+             number of cores. This could be because ~
+             #'QVM:PREPARE-FOR-PARALLELIZATION was called too early. ~
+             The number of workers is ~D and the number of logical ~
+             cores is ~D."
+          num-workers
+          (count-logical-cores))
+  (when (and lparallel:*kernel*
+             (integerp num-workers)
+             (not (= (lparallel:kernel-worker-count) num-workers))) ; force creating a new kernel
```

Comment:
`(not (=` is `(/=`

also prefer the non-parenthesized thing in front of the parenthesized: `(/= num-workers (lparallel:kernel-worker-count ...`

Reply:
> One other thing I would like for you to ensure is that we aren't calling `prepare-for-parallelization` so frequently. I don't think we are, but it would be worth checking before merging.

Ok. The author of lparallel is (rightly) concerned about frequent (or any really) reallocation of the thread pool. It is really easy to write code (in other languages) that thrashes because most time is spent creating and destroying threads. I don't have a good sense at the moment about how exposed this function is, how it might be misused.  The danger is partly mitigated by the fact that nothing expensive is done if the number of requested workers does not change:
```lisp
CL-USER> (time (dotimes (i (expt 10 8)) (qvm:prepare-for-parallelization 2)))
Evaluation took:
  0.988 seconds of real time
  0.988120 seconds of total run time (0.988044 user, 0.000076 system)
  100.00% CPU
  2,277,573,376 processor cycles
  0 bytes consed
```

I also benchmarked `(count-logical-cores)`. It takes about 20ns on my machine. That's why I opted not to capture its return value with `let`.
```lisp
CL-USER> (time (dotimes (i (expt 10 7)) (qvm::count-logical-cores)))
Evaluation took:
  0.210 seconds of real time
  0.209258 seconds of total run time (0.209258 user, 0.000000 system)
```
The main motivation for this PR is to facilitate benchmarking and tuning. Other options for benchmarking are:  1) a bash script and the compiled qvm. and 2) writing additional functions to control the number of threads, just for testing. But, number 2 would probably introduce more complexity and bugs.